salt/modules/cmdmod.py

```diff
@@ -407,16 +407,27 @@ def _get_stripped(cmd):
         return win_runas(cmd, runas, password, cwd)
 
     if runas and salt.utils.platform.is_darwin():
-        # we need to insert the user simulation into the command itself and not
-        # just run it from the environment on macOS as that
-        # method doesn't work properly when run as root for certain commands.
+        # We need to insert the user simulation into the command itself and not
+        # just run it from the environment on macOS as that method doesn't work
+        # properly when run as root for certain commands.
         if isinstance(cmd, (list, tuple)):
             cmd = ' '.join(map(_cmd_quote, cmd))
 
-        cmd = 'su -l {0} -c "{1}"'.format(runas, cmd)
-        # set runas to None, because if you try to run `su -l` as well as
-        # simulate the environment macOS will prompt for the password of the
-        # user and will cause salt to hang.
+        # Ensure directory is correct before running command
+        cmd = 'cd -- {dir} && {{ {cmd}\n }}'.format(dir=_cmd_quote(cwd), cmd=cmd)
+
+        # Ensure environment is correct for a newly logged-in user by running
+        # the command under bash as a login shell
+        cmd = '/bin/bash -l -c {cmd}'.format(cmd=_cmd_quote(cmd))
+
+        # Ensure the login is simulated correctly (note: su runs sh, not bash,
+        # which causes the environment to be initialised incorrectly, which is
+        # fixed by the previous line of code)
+        cmd = 'su -l {0} -c {1}'.format(_cmd_quote(runas), _cmd_quote(cmd))
```

Comment:
cmd = [ 'su', '-l', runas, '-c', 'cd \'{0}\' && {1}'.format(_cmd_quote(cwd), cmd) ] 
When past as a list a shell is **not** invoke to before running the su (no security issue).  When past as a string a shell is invoke to call the su and its arguments which means $abc is expanded before calling su and at the high privileged.(security issue).

Reply:
The command is coerced to a string before executing, so either case is the same. I have covered this specific exploit in the integration test  `test_avoid_injecting_shell_code_as_root` (part of this PR) which passes on the machines I have tested on. Do you know of an input which could bypass the current quoting mechanisms?